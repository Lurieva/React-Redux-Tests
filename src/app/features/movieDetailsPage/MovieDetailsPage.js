import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';

import { InfoPanel, Movies, Header } from '../../components';
import MovieDetails from './movieDetails/MovieDetails';
import GenresInfo from './genresInfo/GenresInfo';
import { BASE_URL } from '../../app.config';

import { receiveMovies, receiveMovie } from '../../../actions/actions';
import { getMoviesByGenre } from '../../../selectors/index';

const mapStateToProps = (state) => ({
    movies: getMoviesByGenre(state),
    movie: state.movie
});
  
const mapDispatchToProps = {
    receiveMovies,
    receiveMovie
};

class MovieDetailsPage extends Component {
    componentDidMount () {
        const movieId = this.props.match.params.movieId;
          
        this.getMovieDetail(movieId);
        
        if (!this.props.movies.length) {
            this.getMovies();
        }
    }

    componentWillReceiveProps(nextProps) {
        const nextMovieId = nextProps.match.params.movieId;

        if (this.props.match.params.movieId !== nextMovieId) {
            this.getMovieDetail(nextMovieId);
        }
    }

    getMovieDetail(movieId) {
        const movie = fetch(`${BASE_URL}movies/${movieId}`)
            .then((res) => res.json())
            .then((movie) => this.props.receiveMovie(movie));
    }

    getMovies() {
        const movies = fetch(`${BASE_URL}movies`)
            .then((res) => res.json())
            .then((res) => this.props.receiveMovies(res.data));
    }

    render() {
        const { movie, movies } = this.props;

        return (
            <Fragment>
                <Header isShowSearchBtn={true}>
                    <MovieDetails movie={movie} />
                </Header>
                <InfoPanel>
                    <GenresInfo genres={movie.genres}/>
                </InfoPanel>
                <Movies movies={movies} />
            </Fragment>
        )
    }
};

export default connect(
    mapStateToProps,
    mapDispatchToProps,
)(MovieDetailsPage);