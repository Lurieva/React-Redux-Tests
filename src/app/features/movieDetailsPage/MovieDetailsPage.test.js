import React from 'react';
import { shallow } from 'enzyme';
import MovieDetailsPage from './MovieDetailsPage';

jest.mock('../../components/header/Header', () => 'Header');
jest.mock('../../components/infoPanel/InfoPanel', () => 'InfoPanel');
jest.mock('../../components/movies/Movies', () => 'Movies');
jest.mock('../../components/sortBy/SortBy', () => 'SortBy');
jest.mock('./movieDetails/MovieDetails', () => 'MovieDetails');
jest.mock('./genresInfo/GenresInfo', () => 'GenresInfo');

xdescribe('MovieDetailsPage', () => {
    let tree;
    let matchMock = {
        params: {
            movieId: 2
        }
    };
    const moviesMock = [
        {
            id: 1,
            title: 'title_1',
            date: '12-01-2018'
        }, {
            id: 2,
            title: 'title_2',
            date: '13-01-2018'
        }
    ];

    beforeEach(() => {
        global.fetch = jest.fn().mockImplementation(() => {
            const p = new Promise((resolve, reject) => {
                resolve({
                    ok: true, 
                    json: () => { 
                        return {
                            data: moviesMock
                        };
                    }
                });
            });
            
            return p;
        });
    });

    beforeEach(() => {
        tree = shallow(<MovieDetailsPage match={matchMock}/>);
    });

    it('should be defined', () => {
        expect(MovieDetailsPage).toBeDefined();
    });

    it('should render correctly ', () => {
        expect(tree).toMatchSnapshot();
    });

    it('should match interface', () => {
        expect(tree.instance().componentDidMount).toBeDefined();
        expect(tree.instance().componentWillReceiveProps).toBeDefined();
        expect(tree.instance().getMovies).toBeDefined();
        expect(tree.instance().getMovieDetail).toBeDefined();
        expect(tree.instance().render).toBeDefined();
    });

    it('should call fetch when component did mount', () => {
        tree.instance().props = {
            match: {
                params: {
                    movieId: 2
                }
            }
        };
        tree.instance().componentDidMount();

        expect(global.fetch).toHaveBeenCalledWith('http://react-cdp-api.herokuapp.com/movies/2');
        expect(global.fetch).toHaveBeenCalledWith('http://react-cdp-api.herokuapp.com/movies');
    });

    it('should call fetch', () => {
        const props = {
            match: {
                params: {
                    movieId: 3
                }
            }
        };

        tree.instance().props = {
            match: {
                params: {
                    movieId: 2
                }
            }
        };
        tree.instance().componentWillReceiveProps(props);
        expect(global.fetch).toHaveBeenCalledWith('http://react-cdp-api.herokuapp.com/movies/2');
    });   
});
