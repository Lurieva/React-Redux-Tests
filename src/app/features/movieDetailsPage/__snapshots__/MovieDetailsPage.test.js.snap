// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieDetailsPage should render correctly  1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <MovieDetailsPage
    match={
      Object {
        "params": Object {
          "movieId": 2,
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Header
          isShowSearchBtn={true}
        >
          <MovieDetails
            movie={Object {}}
          />
        </Header>,
        <InfoPanel>
          <GenresInfo
            genres={undefined}
          />
        </InfoPanel>,
        <Movies
          movies={Array []}
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <MovieDetails
            movie={Object {}}
          />,
          "isShowSearchBtn": true,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "movie": Object {},
          },
          "ref": null,
          "rendered": null,
          "type": "MovieDetails",
        },
        "type": "Header",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <GenresInfo
            genres={undefined}
          />,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "genres": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": "GenresInfo",
        },
        "type": "InfoPanel",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "movies": Array [],
        },
        "ref": null,
        "rendered": null,
        "type": "Movies",
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Header
            isShowSearchBtn={true}
          >
            <MovieDetails
              movie={Object {}}
            />
          </Header>,
          <InfoPanel>
            <GenresInfo
              genres={undefined}
            />
          </InfoPanel>,
          <Movies
            movies={Array []}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <MovieDetails
              movie={Object {}}
            />,
            "isShowSearchBtn": true,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "movie": Object {},
            },
            "ref": null,
            "rendered": null,
            "type": "MovieDetails",
          },
          "type": "Header",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <GenresInfo
              genres={undefined}
            />,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "genres": undefined,
            },
            "ref": null,
            "rendered": null,
            "type": "GenresInfo",
          },
          "type": "InfoPanel",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "movies": Array [],
          },
          "ref": null,
          "rendered": null,
          "type": "Movies",
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
